better error message when remote end disconnects

in #6245, we see the remote end disconnecting. The log looks like

```
2018/05/07 17:24:43 packer.exe: 2018/05/07 17:24:43 [ERROR] Remote command exited without exit status or exit signal.
2018/05/07 17:24:43 packer.exe: 2018/05/07 17:24:43 [INFO] RPC endpoint: Communicator ended with: 2300218
2018/05/07 17:24:43 [INFO] 291 bytes written for 'stderr'
2018/05/07 17:24:43 [INFO] 107 bytes written for 'stdout'
2018/05/07 17:24:43 [INFO] RPC client: Communicator ended with: 2300218
2018/05/07 17:24:43 [INFO] RPC endpoint: Communicator ended with: 2300218
2018/05/07 17:24:43 packer.exe: 2018/05/07 17:24:43 [INFO] 291 bytes written for 'stderr'
2018/05/07 17:24:43 packer.exe: 2018/05/07 17:24:43 [INFO] 107 bytes written for 'stdout'
2018/05/07 17:24:43 packer.exe: 2018/05/07 17:24:43 [INFO] RPC client: Communicator ended with: 2300218
2018/05/07 17:24:43 [INFO] (telemetry) ending shell
2018/05/07 17:24:43 ui error: ==> virtualbox-iso: Step "StepProvision" failed, aborting...
2018/05/07 17:24:43 [INFO] (telemetry) ending virtualbox-iso
2018/05/07 17:24:43 ui error: Build 'virtualbox-iso' errored: unexpected EOF
```

We see the error log from the communicator, `Remote command exited without exit status or exit signal.`, and if we look to the shell provisioner, we see it should be returning a more informative error message: https://github.com/hashicorp/packer/blob/8848682c358bb31597d54a57f196face92cd86df/provisioner/shell/provisioner.go#L284-L290

The more informative message is not being displayed because packer was invoked with `-on-error=abort`. For some reason this results in an "unexpected EOF" error rather than the actual error that was returned by the provisioner.

--- a/common/multistep_runner.go
+++ b/common/multistep_runner.go
@@ -73,7 +73,7 @@ func (s abortStep) Run(ctx context.Context, state multistep.StateBag) multistep.
 func (s abortStep) Cleanup(state multistep.StateBag) {
 	err, ok := state.GetOk("error")
 	if ok {
-		s.ui.Error(err.(error).Error())
+		s.ui.Error(fmt.Sprintf("%s", err))
 	}
 	if _, ok := state.GetOk(multistep.StateCancelled); ok {
 		s.ui.Error("Interrupted, aborting...")
@@ -105,7 +105,7 @@ func (s askStep) Run(ctx context.Context, state multistep.StateBag) (action mult
 
 		err, ok := state.GetOk("error")
 		if ok {
-			s.ui.Error(err.(error).Error())
+			s.ui.Error(fmt.Sprintf("%s", err))
 		}
 
 		switch ask(s.ui, typeName(s.step), state) {

--- a/common/multistep_runner.go
+++ b/common/multistep_runner.go
@@ -71,6 +71,10 @@ func (s abortStep) Run(ctx context.Context, state multistep.StateBag) multistep.
 }
 
 func (s abortStep) Cleanup(state multistep.StateBag) {
+	err, ok := state.GetOk("error")
+	if ok {
+		s.ui.Error(err.(error).Error())
+	}
 	if _, ok := state.GetOk(multistep.StateCancelled); ok {
 		s.ui.Error("Interrupted, aborting...")
 		os.Exit(1)
@@ -97,6 +101,11 @@ func (s askStep) Run(ctx context.Context, state multistep.StateBag) (action mult
 
 		if action != multistep.ActionHalt {
 			return
+		}
+
+		err, ok := state.GetOk("error")
+		if ok {
+			s.ui.Error(err.(error).Error())
 		}
 
 		switch ask(s.ui, typeName(s.step), state) {

--- a/communicator/ssh/communicator.go
+++ b/communicator/ssh/communicator.go
@@ -335,7 +335,6 @@ func (c *comm) reconnect() (err error) {
 	}
 
 	if err != nil {
-		log.Printf("[ERROR] handshake error: %s", err)
 		return
 	}
 	log.Printf("[DEBUG] handshake complete!")

